trieste/utils/misc.py

```diff
@@ -220,21 +233,84 @@ def map_values(f: Callable[[U], V], mapping: Mapping[K, U]) -> Mapping[K, V]:
 """ An unbound type variable. """
 
 
-def get_value_for_tag(mapping: Optional[Mapping[Tag, T]], tag: Tag = OBJECTIVE) -> Optional[T]:
-    """Return the value of a tag in a mapping.
+def get_value_for_tag(
+    mapping: Optional[Mapping[Tag, T]], tags: Union[Tag, Sequence[Tag]] = OBJECTIVE
+) -> Tuple[Optional[Tag], Optional[T]]:
+    """Return the value from a mapping for the first tag found from a sequence of tags.
 
     :param mapping: A mapping from tags to values.
-    :param tag: A tag.
-    :return: The value of the tag in the mapping, or None if the mapping is None.
-    :raises ValueError: If the tag is not in the mapping and the mapping is not None.
+    :param tags: A tag or a sequence of tags. Sequence is searched in order.
+    :return: The chosen tag and value of the tag in the mapping, or None for each if the mapping is
+        None.
+    :raises ValueError: If none of the tags are in the mapping and the mapping is not None.
     """
 
+    if isinstance(tags, Tag):
+        tags = [tags]
+
     if mapping is None:
-        return None
-    elif tag in mapping.keys():
-        return mapping[tag]
+        return None, None
     else:
-        raise ValueError(f"tag '{tag}' not found in mapping")
+        matched_tags = sorted(set(tags) & set(mapping.keys()), key=tags.index)
+        if matched_tags:
+            return matched_tags[0], mapping[matched_tags[0]]
+        else:
+            raise ValueError(f"none of the tags '{tags}' found in mapping")
+
+
+@dataclass(frozen=True)
+class LocalTag:
+    """Manage a tag for a local model or dataset."""
+
+    global_tag: Tag
+    local_index: Optional[int]
+
+    def __post_init__(self) -> None:
+        if self.local_index is not None and self.local_index < 0:
+            raise ValueError(f"local index must be non-negative, got {self.local_index}")
+
+    @property
+    def is_local(self) -> bool:
+        """Return True if the tag is a local tag."""
+        return self.local_index is not None
+
+    @property
+    def tag(self) -> Tag:
+        """The local tag."""
+        if self.is_local:
+            return f"{self.global_tag}__{self.local_index}"
+        else:
+            return self.global_tag
+
+    def __repr__(self) -> str:
+        """Return the local tag."""
+        return f"LocalTag({self.global_tag}, {self.local_index})"
+
+    def __str__(self) -> str:
+        """Return the local tag."""
+        return str(self.tag)
+
+    def __hash__(self) -> int:
+        """Return the hash of the overall tag."""
+        return hash(self.tag)
+
+    def __eq__(self, other: object) -> bool:
+        """Return True if the local tag is equal to the other object."""
+        return hash(self) == hash(other)
+
+    @staticmethod
+    def from_tag(tag: Union[Tag, LocalTag]) -> LocalTag:
+        """Return a LocalTag from a given tag."""
+        if isinstance(tag, LocalTag):
+            return tag
+        else:
+            tag = str(tag)
```

Comment:
This will become much simpler if we use LocalTags as tags unchanged! (Though if that doesn't work out then I might have some other comments...)